packages/utilities/src/lib/sleep.ts

```diff
@@ -0,0 +1,65 @@
+export interface SleepOptions {
+	/**
+	 * When provided the corresponding `AbortController` can be used to cancel an asynchronous action.
+	 */
+	signal?: AbortSignal | undefined;
+
+	/**
+	 * Set to `false` to indicate that the scheduled `Timeout`
+	 * should not require the Node.js event loop to remain active.
+	 * @default true
+	 */
+	ref?: boolean | undefined;
+}
+
+export class AbortError extends Error {
+	public code: string;
+	public constructor(
+		message?: string,
+		options?: {
+			cause?: unknown;
+			code?: string;
+		}
+	) {
+		super(message, options);
+		this.name = 'AbortError';
+		this.code = options?.code ?? 'ERR_ABORT';
+	}
+}
+
+/**
+ * Sleeps for the specified number of milliseconds.
+ * @param ms The number of milliseconds to sleep.
+ * @param value A value with which the promise is fulfilled.
+ * @see {@link sleepSync} for a synchronous version.
+ */
+export function sleep<T = undefined>(ms: number, value?: T, options?: SleepOptions): Promise<T> {
+	return new Promise((resolve, reject) => {
+		const timer: NodeJS.Timeout | number = setTimeout(() => resolve(value!), ms);
+		if (options?.signal) {
+			options.signal.addEventListener('abort', () => {
+				clearTimeout(timer);
+				reject(
+					new AbortError('The operation was aborted', {
+						cause: options.signal!.reason
```

Comment:
You need to assign the signal somewhere, the reason why TS thinks it's nullable is because it's *potentially nullable*.

```ts
const options = { signal: controller.signal };
const promise = sleep(1000, undefined, options);
options.signal = undefined;
controller.abort();
await promise; // <-- ReferenceError!
```

To fix this, assign `options.signal` to a local variable.